compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/LinalgOpInfo.h"
+
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+
+using namespace mlir::linalg;
+
+namespace mlir {
+namespace iree_compiler {
+
+LinalgOpInfo::LinalgOpInfo(linalg::LinalgOp linalgOp) { computeInfo(linalgOp); }
+
+/// Returns true if `map` is a tranposition.
+// TODO(dcaballe): Discern between "memcopy" transposes and "shuffle"
+// transposes.
+// TODO(dcaballe): Shouldn't there be a utility for this somewhere in Affine?
```

Comment:
Are you looking for `map.isProjectedPermutation()` or `map.isPermutation()`?

I don't really follow the implementation. It looks like checking if there is a pair `(d_i, d_{i+1})` where `d_i > d_{i+1}`. Shouldn't we check if they are unique?

Reply:
The projected permutation check is not enough because (d0, d1, d2) -> (d0, d2) is a projected permutation but not a transpose. We also have to take into account broadcasts. I guess we could check that the map is a projected permutation allowing zeros in the result and then add an extra check to make sure that there exist `i` and `j` in the map input such that di < dj and di > dj in the map output. Perhaps what I'm describing is the negative of `isMinorIdentityMapWithBroadcast`? I'll give it a try

Reply:
Added doc and improve the check but I can't find a check that does what we need here in affine map utils.